src/rnnlm/arpa-sampling.cc

```diff
@@ -0,0 +1,393 @@
+        size_ngrams++; // number of words given
+      }
+    }
+    KALDI_LOG << "Read in number of " << (i + 1) << "-grams: " << size_ngrams;
+  }
+  KALDI_LOG << "Assert sum of unigram probs equal to 1...";
+  BaseFloat prob_sum = 0.0;
+  int32 count = 0;
+  for (auto it1 = (probs_[0]).begin(); it1 != (probs_[0]).end();++it1) {
+    HistType h(it1->first);
+    for (auto it2 = (probs_[0])[h].begin(); it2 != (probs_[0])[h].end(); ++it2) {
+      prob_sum += 1.0 * pow(10.0, (it2->second).first);
+      count++;
+    }
+  }
+  KALDI_LOG << "Number of total words: " << count;
+  KALDI_LOG << "Sum of unigram probs equal to " << prob_sum;
+
+  KALDI_LOG << "Assert sum of bigram probs given a history equal to 1...";
+  prob_sum = 0.0;
+  auto it1 = probs_[1].begin();
+  HistType h(it1->first);
+  for (int32 i = 0; i < num_words_; i++) {
+    auto it2 = probs_[1][h].find(vocab_[i].second);
+    if (it2 != probs_[1][h].end()) {
+      prob_sum += 1.0 * pow(10, (it2->second).first);
+    } else {
+      prob_sum += pow(10, GetProb(2, vocab_[i].second, h));
+    }
+  }
+  KALDI_LOG << "Sum of bigram probs given a history equal to " << prob_sum;
+}
+
+// Read sentences from a file
+void ArpaSampling::ReadSentences(std::istream &iss, std::vector<std::vector<int32> >* sentences) {
+  const fst::SymbolTable* sym = Symbols();
+  std::string line;
+  KALDI_LOG << "Start reading sentences...";
+  while (getline(iss, line)) {
+    std::istringstream is(line);
+    std::istream_iterator<std::string> begin(is), end;
+    std::vector<std::string> tokens(begin, end);
+    std::vector<int32> sentence;
+    int32 word;
+    int32 bos = sym->Find(bos_symbol_);
+    sentence.push_back(bos);
+    for (int32 i = 0; i < tokens.size(); i++) {
+      word = sym->Find(tokens[i]);
+      if (word == fst::SymbolTable::kNoSymbol) {
+        word = sym->Find(unk_symbol_);
+      }
+      sentence.push_back(word);
+    }
+    int32 eos = sym->Find(eos_symbol_);
+    sentence.push_back(eos);
+    (*sentences).push_back(sentence);
+  }
+  KALDI_LOG << "Finished reading sentences.";
+}
+
+// Read histories of integers from a file
+void ArpaSampling::ReadHistories(std::istream &is, bool binary) {
+  if (binary) {
+    KALDI_ERR << "binary-mode reading is not implemented for ArpaFileParser";
+  }
+  const fst::SymbolTable* sym = Symbols();
+  std::string line;
+  KALDI_LOG << "Start reading histories from file...";
+  while (getline(is, line)) {
+    std::istringstream is(line);
+    std::istream_iterator<std::string> begin(is), end;
+    std::vector<std::string> tokens(begin, end);
+    HistType history;
+    int32 word;
+    for (int32 i = 0; i < tokens.size(); i++) {
+      word = sym->Find(tokens[i]);
+      if (word == fst::SymbolTable::kNoSymbol) {
+        word = sym->Find(unk_symbol_);
+      }
+      history.push_back(word);
+    }
+    if (history.size() >= ngram_order_) {
+      std::reverse(history.begin(), history.end());
```

Comment:
this is an extremely inefficient way of doing things. please make it more efficient.